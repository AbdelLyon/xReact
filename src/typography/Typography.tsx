import { forwardRef, PropsWithChildren } from "react";

import { cn } from "@/utils";

type TitleProps = {
  title: string;
  isLoading?: boolean;
  className?: string;
  loading?: () => React.ReactNode;
  tag?: Extract<
    keyof JSX.IntrinsicElements,
    "h1" | "h2" | "h3" | "h4" | "h5" | "h6"
  >;
};

const Title = forwardRef<HTMLHeadingElement, TitleProps>(
  (
    {
      title,
      isLoading = false,
      tag: CustomTitle = "h1",
      loading,
      className,
      ...props
    },
    ref,
  ) => {
    if (isLoading && loading) {
      return loading();
    }

    return (
      <CustomTitle
        ref={ref}
        className={cn(
          "text-xl scroll-m-20 font-bold tracking-tight",
          className,
        )}
        {...props}
      >
        {title}
      </CustomTitle>
    );
  },
);

Title.displayName = "Title";

type BlockquoteProps = PropsWithChildren<{
  className?: string;
}>;

const Blockquote = forwardRef<HTMLQuoteElement, BlockquoteProps>(
  ({ className, children, ...props }, ref) => (
    <blockquote
      ref={ref}
      className={cn("mt-6 border-l-2 pl-6 italic", className)}
      {...props}
    >
      {children}
    </blockquote>
  ),
);

Blockquote.displayName = "Blockquote";

type InlineCodeProps = PropsWithChildren<{
  className?: string;
}>;

const InlineCode = forwardRef<HTMLElement, InlineCodeProps>(
  ({ className, children, ...props }, ref) => (
    <code
      ref={ref}
      className={cn(
        "bg-muted relative rounded px-[0.3rem] py-[0.2rem] font-mono text-sm font-semibold",
        className,
      )}
      {...props}
    >
      {children}
    </code>
  ),
);

InlineCode.displayName = "InlineCode";

export { Title, Blockquote, InlineCode };
