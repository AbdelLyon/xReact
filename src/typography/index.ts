import "@/index.css";
export { Title, Blockquote, InlineCode } from "@/typography/Typography";
