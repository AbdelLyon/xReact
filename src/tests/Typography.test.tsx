import { render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import { Blockquote, InlineCode, Title } from "@/typography";

describe("Title component", () => {
  const loadingMock = vi.fn(() => <div>Loading...</div>);
  const titleText = "Test Title";

  it("should match snapshot when isLoading is false", () => {
    const { asFragment } = render(
      <Title loading={() => <p>...chargement</p>} title={titleText} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match snapshot when isLoading is true", () => {
    const { asFragment } = render(
      <Title isLoading={true} loading={loadingMock} title={titleText} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe("Blockquote component", () => {
  it("should match snapshot with given className", () => {
    const { asFragment } = render(
      <Blockquote className="extra-class">{'"Quote"'}</Blockquote>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe("InlineCode component", () => {
  it("should match snapshot with given className", () => {
    const { asFragment } = render(
      <InlineCode className="code-extra">code</InlineCode>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
