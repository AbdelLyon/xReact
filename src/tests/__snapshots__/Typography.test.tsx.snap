// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Blockquote component > should match snapshot with given className 1`] = `
<DocumentFragment>
  <blockquote
    class="mt-6 border-l-2 pl-6 italic extra-class"
  >
    "Quote"
  </blockquote>
</DocumentFragment>
`;

exports[`InlineCode component > should match snapshot with given className 1`] = `
<DocumentFragment>
  <code
    class="bg-muted relative rounded px-[0.3rem] py-[0.2rem] font-mono text-sm font-semibold code-extra"
  >
    code
  </code>
</DocumentFragment>
`;

exports[`Title component > should match snapshot when isLoading is false 1`] = `
<DocumentFragment>
  <h1
    class="text-xl scroll-m-20 font-bold tracking-tight"
  >
    Test Title
  </h1>
</DocumentFragment>
`;

exports[`Title component > should match snapshot when isLoading is true 1`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;
